### PR TITLE
Meter mode code refactor

### PR DESCRIPTION
**Pull request by Explorer09**
The main goal of this set of patches is to remove the use of `CUSTOM_METERMODE` in `defaultMode` values, and to adjust `Meter_setMode()` code in preparation of the `supportedModes` feature (PR #1387).

### CODE REVIEW

**Comment by BenBE** [comment]
Any reason, why you dropped 25388969 from this series?

**Comment by Explorer09** [comment]
@BenBE It's better for 25388969d4df6a123ddac561a66a37188837c738 to merge with your PR and that change is slightly irrelevant to the other parts of code refactoring here. Besides the function name `nextSupportedMode` would make less sense until the real `supportedMode` field is implemented.

**Comment by BenBE** [comment]
Sure, just noticed when comparing your branches.

LGTM.

**Comment by Explorer09** [comment]
More code refactoring, and thus it's good to re-review this.
* Make `Meter_modes` array private to `Meter.c`
* Define type `MeterModeId` to `uint8_t` and make a separate header for it. (If I put that `typedef` in `Meter.h` it could cause a header dependency hell that I had a hard time resolving it.)

File: Settings.c [hunk 1]
```diff
@@ -605,7 +605,7 @@ static void writeMeterModes(const Settings* this, FILE* fd, char separator, unsi
    if (this->hColumns[column].len) {
       const char* sep = "";
       for (size_t i = 0; i < this->hColumns[column].len; i++) {
-         fprintf(fd, "%s%d", sep, this->hColumns[column].modes[i]);
+         fprintf(fd, "%s%d", sep, (int)this->hColumns[column].modes[i]);
```
Comment (BenBE) [inline Settings.c:608]: Avoid this cast by having `MeterModeId` be an alias for `int`; which is also basically what the enum would have been; `uint8_t` doesn't save anything, as inside the structs where it's used it causes padding to be used thus negating any gains …

Comment (Explorer09) [reply Settings.c:608]: @BenBE 
Um... This use of `MeterModeId` is an array in `MeterColumnSetting`. Not part of `Meter` structures.

Comment (BenBE) [reply Settings.c:608]: Sure. I was referring to the use as part of `struct Meter_`.

Also, the format string is `%d` in this line; thus forcing your cast.

Comment (Explorer09) [reply Settings.c:608]: I don't know. Is the `MeterModeId` array in `MeterColumnSetting` meant to have a short lifetime? If that's the case, then allocating a bigger `int` array wouldn't hurt.

(Note that in a variadic function like `printf`, any argument shorter than `int` would implicitly promote to `int` anyway even without the cast.)

**Comment by BenBE** [comment]
* In e5bcc4fe there's unrelated changes in `Header.c`.
* In e5bcc4fe for `MeterMode.h` do `typedef int MeterModeId;` (saving on including `stdint.h` and several casts)
* In e5bcc4fe for `Settings.c` inside `Settings_defaultMeters` are unrelated changes
* In dab8ebf7 in `MeterMode.h` consider `typedef enum MeterModeId_ { … } MeterModeId;`
* In dab8ebf7 in `MeterMode.h` there should be two blank lines before declarations
* In 2461e6e2 changes for code structure and functional changes are mixed

**Comment by Explorer09** [comment]
@BenBE 
> * In dab8ebf7f94e868cfa22ddda47e190645ef38825 in `MeterMode.h` there should be two blank lines before declarations

Will do.

> * In e5bcc4fecac5b0364bc64bfd3f4a127217ed8513 there's unrelated changes in `Header.c`.
> * In 2461e6e2b6b5324a775cdabef5dcbdd7d7192a5e changes for code structure and functional changes are mixed
> * In e5bcc4fecac5b0364bc64bfd3f4a127217ed8513 for `Settings.c` inside `Settings_defaultMeters` are unrelated changes

All of these are style changes being mixed with changes in behavior. Do you mean to drop the style changes or split them into seperate commits?

> * In e5bcc4fecac5b0364bc64bfd3f4a127217ed8513 for `MeterMode.h` do `typedef int MeterModeId;` (saving on including `stdint.h` and several casts)
> * In dab8ebf7f94e868cfa22ddda47e190645ef38825 in `MeterMode.h` consider `typedef enum MeterModeId_ { … } MeterModeId;`
> 

I think you're interested in some size reports:

```text
gcc-11 with options -O2 -flto -Wl,--as-needed

    text    data     bss     dec    hex
  359500   26248    6520  392268  5fc4c (a) MeterModeId in uint8_t
  359460   26248    6520  392228  5fc24 (b) MeterModeId in signed int
  359452   26248    6520  392220  5fc1c (c) MeterModeId in unsigned int
  359500   26248    6520  392268  5fc4c (d) uint8_t MeterModeId with reorder
                                            of Meter & MeterClass structures

gcc-11 with options -Os -flto -Wl,--as-needed

    text    data     bss     dec    hex
  282899   26320    6552  315771  4d17b (a) MeterModeId in uint8_t
  282867   26320    6552  315739  4d15b (b) MeterModeId in signed int
  282865   26320    6552  315737  4d159 (c) MeterModeId in unsigned int
  282899   26320    6552  315771  4d17b (d) uint8_t MeterModeId with reorder 
                                            of Meter & MeterClass structures

clang-16 with options -Os -flto -Wl,--as-needed

    text    data     bss     dec    hex
  291703   25380    6368  323451  4ef7b (a) MeterModeId in uint8_t
  291693   25380    6368  323441  4ef71 (b) MeterModeId in signed int
  291692   25380    6368  323440  4ef70 (c) MeterModeId in unsigned int
  291703   25004    6368  323075  4ee03 (d) uint8_t MeterModeId with reorder
                                            of Meter & MeterClass structures
```
I gathered these from GitHub Actions CI. And you are right in that using `uint8_t` for `MeterModeId` doesn't help with the code size, but interestingly making `MeterModeId` an `unsigned int` (the (c) case) can reduce a few bytes of the compiled binary.
The (d) case is interesting and I didn't figure out the cause yet (I moved the `mode` member to close to the `maxItems` and `curItems` members of the `MeterClass` and `Meter` structures, thus kills the unnecessary padding). I think I'll pick the (c) case.

**Comment by Explorer09** [comment]
> * In 2461e6e2b6b5324a775cdabef5dcbdd7d7192a5e changes for code structure and functional changes are mixed

I'm a little confused what you mean by "code structure and functional changes are mixed" and how should I split the commits.

**Comment by BenBE** [comment]
> @BenBE
> 
> > * In [e5bcc4f](https://github.com/htop-dev/htop/commit/e5bcc4fecac5b0364bc64bfd3f4a127217ed8513) there's unrelated changes in `Header.c`.
> > * In [2461e6e](https://github.com/htop-dev/htop/commit/2461e6e2b6b5324a775cdabef5dcbdd7d7192a5e) changes for code structure and functional changes are mixed
> > * In [e5bcc4f](https://github.com/htop-dev/htop/commit/e5bcc4fecac5b0364bc64bfd3f4a127217ed8513) for `Settings.c` inside `Settings_defaultMeters` are unrelated changes
> 
> All of these are style changes being mixed with changes in behavior. Do you mean to drop the style changes or split them into seperate commits?

In the first commit, there's this `sizeof` change in the commit, that's not purely code formatting, but a distinct change (although no functional difference). Similar goes for the other changes I listed. Just split those changes (you can merge the two sizeof changes in one commit afterwards).

> > * In [e5bcc4f](https://github.com/htop-dev/htop/commit/e5bcc4fecac5b0364bc64bfd3f4a127217ed8513) for `MeterMode.h` do `typedef int MeterModeId;` (saving on including `stdint.h` and several casts)
> > * In [dab8ebf](https://github.com/htop-dev/htop/commit/dab8ebf7f94e868cfa22ddda47e190645ef38825) in `MeterMode.h` consider `typedef enum MeterModeId_ { … } MeterModeId;`
> 
> I think you're interested in some size reports:
> 
> ```
> gcc-11 with options -O2 -flto -Wl,--as-needed
> 
>     text    data     bss     dec    hex
>   359500   26248    6520  392268  5fc4c (a) MeterModeId in uint8_t
>   359460   26248    6520  392228  5fc24 (b) MeterModeId in signed int
>   359452   26248    6520  392220  5fc1c (c) MeterModeId in unsigned int
>   359500   26248    6520  392268  5fc4c (d) uint8_t MeterModeId with reorder
>                                             of Meter & MeterClass structures
> 
> gcc-11 with options -Os -flto -Wl,--as-needed
> 
>     text    data     bss     dec    hex
>   282899   26320    6552  315771  4d17b (a) MeterModeId in uint8_t
>   282867   26320    6552  315739  4d15b (b) MeterModeId in signed int
>   282865   26320    6552  315737  4d159 (c) MeterModeId in unsigned int
>   282899   26320    6552  315771  4d17b (d) uint8_t MeterModeId with reorder 
>                                             of Meter & MeterClass structures
> 
> clang-16 with options -Os -flto -Wl,--as-needed
> 
>     text    data     bss     dec    hex
>   291703   25380    6368  323451  4ef7b (a) MeterModeId in uint8_t
>   291693   25380    6368  323441  4ef71 (b) MeterModeId in signed int
>   291692   25380    6368  323440  4ef70 (c) MeterModeId in unsigned int
>   291703   25004    6368  323075  4ee03 (d) uint8_t MeterModeId with reorder
>                                             of Meter & MeterClass structures
> ```
> 
> I gathered these from GitHub Actions CI. And you are right in that using `uint8_t` for `MeterModeId` doesn't help with the code size, but interestingly making `MeterModeId` an `unsigned int` (the (c) case) can reduce a few bytes of the compiled binary. The (d) case is interesting and I didn't figure out the cause yet (I moved the `mode` member to close to the `maxItems` and `curItems` members of the `MeterClass` and `Meter` structures, thus kills the unnecessary padding). I think I'll pick the (c) case.

(b) and (c) are fine. Basically we have (b) right now. Be sure to update the format strings accordingly and you'll be able to skip that one cast.

**Pull request merged**